README.md
# NTUNHS-Assistant
北護助手

[手機版各頁面DEMO](https://hackmd.io/@a5566qq123/rkAIK72Jd)

[電腦版各頁面DEMO](https://hackmd.io/@a5566qq123/rk-822YJu)

[定時抓取及更新課程查詢資料 NTUNHS-CourseUpdate](https://github.com/Chinlinlee/NTUNHS-CourseUpdate)
## 初始化
```bash=
npm install
npm run initconfig #這個會創建config檔，請先設定
```

## configure

`models/NTUNHS/config.js` 此設定檔用於登入北護入口網更新教學計畫用
```javascript
module.exports  = {
    stuNum : "入口網帳號" , 
    stuPwd : "入口網密碼"
}
```
`config/config.js` 此設定檔用於設定Server的相關設定

```javascipt
module.exports =
{
    HTTPServer:
    {
        "viewsRoot": "./views",
        "httpPort": 30087,
        "host" : "127.0.0.1" ,
        "timeout": 30000
    },
    MONGODB: {
        "host" : "127.0.0.1" , 
        "port" : 27017 , 
        "db" : "My_ntunhs" , 
        "ssl" : false ,
        "username" : "user" , 
        "password" : "password" , 
        "authDB" : "admin"
    }
};
```


## 建置與啟動
### 建置
- 本地設定完畢後，如果要測試課程查詢可以跑下面這條
```bash=
npm run build #創建當學期課程查詢資料用
```

### 啟動
**注意**
- 如果要在localhost測試請將server.js當中`session`的`cookie`的`sameSite`以及`secure`註解掉，並注意不要commit到此註解
```javascript
cookie : {
      maxAge: 60 * 60* 24 * 1000,
      //sameSite: 'none',
      //secure : true
} ,
```
```bash=
node server.js
```
或使用`pm2`(使用cluster模式)
```bash=
pm2 start ecosystem.config.js
```

### 記錄登入的人數
登入時會記錄到`default.log`，並使用`getTrafic.js`定時於`23:58`記錄當天登入的人數
```bash=
pm2 start getTrafic.js --name mycosim-trafic
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/Chinlinlee/NTUNHS-Assistant/compare/v1.0.0...v1.2.0) (2021-10-23)


### Features

* 更改取得Object的方式 ([c7bd7c9](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/c7bd7c963968def0878ff73db49a3b2fb63016a9))
* 更改api login回傳的資料 ([3f475ee](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/3f475eea853cc92ac8bdea783352e6707b2f39e6))
* 相關課程成績分佈圖標題新增`上課班級` ([61a5ea3](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/61a5ea3ba9a7937c42c40fd05ea309a236cab7ed))
* 將學習地圖回傳資料由html轉為Object ([74dcdaf](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/74dcdafe17b6d7fc8f91fa779ba4065529bd5894))
* 將selenium檢查簽核系統的方法分為立即執行及排程版本 ([eec4b7b](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/eec4b7bb26299e8d29e83c14f83b5e594c2bdfb8))
* 新增用學期成績檢查登入狀態的API ([c250328](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/c250328d26b09b86344046f82898bd31c979ba2e))
* 新增dark mode 的card-header樣式 ([caf61b2](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/caf61b2589aedd3f98e6f2b4dfe0a4cf3fbff80e))
* add update Course schedule ([6041a9e](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/6041a9efcd8fa6834d3d4a1411ecbff7c3d7a4e1))
* **storedHistoryScore:** 更改成績分佈的回傳資料以及排序 ([deafd2a](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/deafd2af3e3f8f223f86772737f4ac50539aedd6))


### Bug Fixes

* 修正取得教學評量是否可用的url ([a386f98](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/a386f98e1bfda1928465681449432c5ce487c1f7))
* 修正執行時會直接停止的問題 ([dd778d4](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/dd778d411cde787092a0dc89a15242e14c3e7cd2))
* 修正獲取修課明細的課程名稱 ([f5915cb](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/f5915cb49d147275b65450307608df9cee9f5708))
* add sameSite to solve ionic can't access the api ([9c50621](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/9c506216130b29bccd97fba73810303b016048ef))

## 1.0.0 (2021-06-28)


### Features

* 大課表可點課顯示更多資訊 ([56865aa](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/56865aaf47ab6716e2064e6d9f44540931bbaa9e))
* 在初始化時加入blockUI ([a143d88](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/a143d889dd1d3ff44867cace88c80db2bd6eb719))
* 在查詢後將conn關閉 ([8ffdbbf](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/8ffdbbf6c3e8916316ba8c5c9c3077d14542f61a))
* 更改初始化config內容 ([15de296](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/15de29652ef7a5b7ab74361eda9223e1e48674dc))
* 更改抓取歷年成績方式 ([12c5117](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/12c51175f07ddd2dc6819515cce6e496884204c9))
* 更改版面 ([bbd7f1c](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/bbd7f1c23a8c504384ef90805cdb15bf3545d2bd))
* 更改獲取修課明細function ([695987a](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/695987a972ed33858c97eb0f1f1290e8c26c225b))
* 更改獲取登入資訊方式 ([288e6e8](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/288e6e8b970fa7d9ae20aa17208c61cf06a3c2ac))
* 更改獲取登入資訊方式 ([8f720c4](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/8f720c43a8ba371535980045820b69e25c2642a9))
* 更改獲取登入資訊方式 ([22b8b02](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/22b8b0297675edf9cbc0303e4b30b18a67ef0c60))
* 更改簽核系統登入方式 ([f76bb19](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/f76bb1971131134390292f25e9e8578a06ca539c))
* 更改export function ([fe9f6a4](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/fe9f6a437fb754ab730b0ae1b640d7e28b7168fe))
* 更改mongodb 連線字串 ([4490d91](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/4490d919693e32efec79263b031761484a761775))
* 更改selenium為fetch ([9566e39](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/9566e39b5ae8661afc3260216d1575c9bd41de0e))
* 紀錄當天登入的人數 ([69e2417](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/69e24177adc4d679420e7439466d8274f3b5ca3c))
* 修改登入邏輯 ([acf4567](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/acf45671cc964cc039c3fac8e44fb89f72e8e2b9))
* 修課明細增加課程14碼 ([5b4e712](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/5b4e712be8d0377a6099cb707f3a018dd271d8c2))
* 個人修課明細增加學期選項 ([54f8641](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/54f8641f5d7f0ee737be6dfaf47125f9ff604b75))
* 無成績時顯示無成績 ([b557300](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/b557300e2998c128024fde7672c0c43eb1644157))
* 登入時增加blockUI ([e460135](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/e460135f322c33a8a24b1c1d81a097f56c1d7258))
* 登出時將session清空 ([030b3c5](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/030b3c5b120f35b0ac16153002f69f99ef83820a))
* 新稱相關課程分佈圖 ([8981602](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/89816022eb844cec94cb924d4831d68cd01f0c69))
* 新增"新增預排"按鈕 ([d711c66](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/d711c66160616ee0f0c3c01721f930d2b4341e0d))
* 新增上傳排名 ([0e429ce](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/0e429ce7d3f8013cd1dd1fbfad1c5d31be19af0a))
* 新增使用者在學的所有學期 ([6bd64f5](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/6bd64f59065e77eadac6dc3bd1acc24e4e34701f))
* 新增後續給分離的前端用的登入api ([65ca1bc](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/65ca1bc1ca3663c1e70ae59ba899f51b986c784b))
* 新增是否可以使用簽核系統api ([fdf83f9](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/fdf83f90707eaecada7be186a2d33dcd2eb67563))
* 新增從簽核系統拿取排名api ([00bb08a](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/00bb08a8d87bead9eb6f01524669964c95159f3f))
* 新增排程刪除過期session ([1052b2a](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/1052b2a55cd96f914daa271d3b8ee7ca44a2d07e))
* 新增匯出預排課表pdf ([7ccd5cb](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/7ccd5cb7fbcfa1e885a1cad1ff5716a65397fc85))
* 新增課程查詢html描述 ([5f453df](https://github.com/Chinlinlee/NTUNHS-Assistant/commit/5f453dfd991b642a697c2385fa24155bb45c72fc))

addConfig.js
const fs  = require('fs');
const mkdirp = require('mkdirp');

function main () {
    mkdirp.sync('./config');
    let mainConfig = 
`
module.exports =
{
    HTTPServer:
    {
        "viewsRoot": "./views",
        "httpPort": 30087,
        "host" : "127.0.0.1" ,
        "timeout": 30000
    },
    MONGODB: {
        "host" : "127.0.0.1" , 
        "port" : 27017 , 
        "db" : "My_ntunhs" , 
        "ssl" : false ,
        "username" : "user" , 
        "password" : "password" , 
        "authDB" : "admin"
    }
};
`
    fs.writeFileSync('./config/config.js' , mainConfig);
    console.log(`創建./config/config.js成功`);

    
    let NTUNHSConfig = `
module.exports  = {
    stuNum : "入口網帳號" , 
    stuPwd : "入口網密碼"
}
    `
    fs.writeFileSync('./models/NTUNHS/config.js' , NTUNHSConfig);
    console.log("創建./models/NTUNHS/config.js成功");
}
main();